client/components/date-range/index.js

```diff
@@ -0,0 +1,511 @@
+		const date = this.props.moment( val, this.getLocaleDateFormat() );
+
+		const fromDate = this.props.moment( this.state.textInputStartDate, this.getLocaleDateFormat() );
+		const toDate = this.props.moment( this.state.textInputEndDate, this.getLocaleDateFormat() );
+
+		// Check date validity
+		const isValidFrom = this.isValidDate( fromDate );
+		const isValidTo = this.isValidDate( toDate );
+
+		// If either of the date inputs are invalid then revert
+		// to current start/end date from state
+		if ( ! isValidFrom || ! isValidTo ) {
+			this.setState( {
+				textInputStartDate: this.formatDateToLocale( this.state.startDate ),
+				textInputEndDate: this.formatDateToLocale( this.state.endDate ),
+			} );
+		}
+
+		const isSameDate = this.state[ `${ startOrEnd.toLowerCase() }Date` ].isSame( date, 'day' );
+
+		// If the new date in the blurred input is valid
+		// and it's not the same as the existing value
+		if ( this.isValidDate( date ) && ! isSameDate ) {
+			this.onSelectDate( date );
+		}
+	};
+
+	/**
+	 * Converts moment dates to a DateRange
+	 * as required by Day Picker DateUtils
+	 * @param  {MomentJSDate} startDate the start date for the range
+	 * @param  {MomentJSDate} endDate   the end date for the range
+	 * @return {Object}           the date range object
+	 */
+	toDateRange( startDate, endDate ) {
+		return {
+			from: this.momentDateToJsDate( startDate ),
+			to: this.momentDateToJsDate( endDate ),
+		};
+	}
+
+	/**
+	 * Handles selection (only) of new dates persisting
+	 * the values to state. Note that if the user does not
+	 * commit the dates (eg: clicking "Apply") then the `revertDates`
+	 * method is triggered which restores the previous ("stale") dates.
+	 *
+	 * Dates are only persisted via the commitDates method.
+	 *
+	 * @param  {MomentJSDate} date the newly selected date object
+	 */
+	onSelectDate = date => {
+		// DateUtils requires a range object with this shape
+		const range = this.toDateRange( this.state.startDate, this.state.endDate );
+
+		const rawDay = this.momentDateToJsDate( date );
+
+		// Calculate the new Date range
+		const newRange = DateUtils.addDayToRange( rawDay, range );
+
+		// Edge case: Range can sometimes be from: null, to: null
+		if ( ! newRange.from || ! newRange.to ) return;
+
+		// Update state to reflect new date range for
+		// calendar and text inputs
+		this.setState(
+			previousState => {
+				let newState = {
+					startDate: this.nativeDateToMoment( newRange.from ),
+					endDate: this.nativeDateToMoment( newRange.to ),
+					textInputStartDate: this.nativeDateToMoment( newRange.from ).format(
+						this.getLocaleDateFormat()
+					),
+					textInputEndDate: this.nativeDateToMoment( newRange.to ).format(
+						this.getLocaleDateFormat()
+					),
+				};
+
+				// For first date selection only: "cache" previous dates
+				// just in case user doesn't "Apply" and we need to revert
+				// to the original dates
+				if ( ! this.state.staleDatesSaved ) {
```

Comment:
Slightly nitty question here (not a blocker): Should we use `previousState. staleDatesSaved` here for the sake of consistency? Or is there purpose in using `this.state` here?

Reply:
Fixed.